File: android/src/main/kotlin/ru/kardenvan/flutter_yandex_ads_plus/platform_api/ad_event/BasicAdEvent.kt
package ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event

data class BasicAdEvent constructor(
    private val viewUid: String,
    private val type: Type,
    private val parameters: Map<String, Any?>? = null,
): AdEvent() {
    override fun toMap(): Map<String, Any?> {
        return mapOf(
            "view_uid" to viewUid,
            "type" to type.value,
            "parameters" to parameters,
        )
    }

    enum class Type constructor(
        val value: String
    ) {
        ON_AD_LOADED("onAdLoaded"),
        ON_AD_FAILED_TO_LOAD("onAdFailedToLoad"),
        ON_IMPRESSION("onImpression"),
        ON_LEFT_APPLICATION("onLeftApplication"),
        ON_RETURNED_TO_APPLICATION("onReturnedToApplication"),
        ON_AD_CLICKED("onAdClicked");
    }
}

File: ios/Classes/platform_api/ad_event/BasicAdEvent.swift
//
//  BannerAdEvent.swift
//  flutter_yandex_ads_plus
//
//  Created by apple on 30.12.2022.
//

import Foundation

class BasicAdEvent: AdEvent {
    init(
        viewUid: String,
        type: BasicAdEventType,
        parameters: [String: Any?]? = nil
    ) {

        super.init(
            viewUid: viewUid,
            type: type.getValue(),
            parameters: parameters
        )
    }
    
    enum BasicAdEventType {
        case onAdLoaded
        case onAdFailedToLoad
        case onAdClicked
        case onLeftApplication
        case didDismissScreen
        case willPresentScreen
        case onImpression
        
        func getValue() -> String {
            switch self {
                case .onAdLoaded:
                    return "onAdLoaded"
                case .onAdFailedToLoad:
                    return "onAdFailedToLoad"
                case .onAdClicked:
                    return "onAdClicked"
                case .onLeftApplication:
                    return "onLeftApplication"
                case .didDismissScreen:
                    return "didDismissScreen"
                case .willPresentScreen:
                    return "willPresentScreen"
                case .onImpression:
                    return "onImpression"
            }
        }
    }
}
